algorithmic_efficiency/workloads/ogb/ogb_jax/input_pipeline.py

```diff
@@ -0,0 +1,161 @@
+    file_data_rng = None
+    dataset_data_rng = None
+
+  read_config = tfds.ReadConfig(add_tfds_id=True, shuffle_seed=file_data_rng)
+  dataset = tfds.load(
+      'ogbg_molpcba',
+      split=split,
+      shuffle_files=should_shuffle,
+      read_config=read_config,
+      data_dir=data_dir)
+
+  if should_shuffle:
+    dataset = dataset.shuffle(seed=dataset_data_rng, buffer_size=2**15)
+    dataset = dataset.repeat()
+
+  return dataset
+
+
+def _to_jraph(example):
+  """Converts an example graph to jraph.GraphsTuple."""
+  example = jax.tree_map(lambda x: x._numpy(), example)  # pylint: disable=protected-access
+  edge_feat = example['edge_feat']
+  node_feat = example['node_feat']
+  edge_index = example['edge_index']
+  labels = example['labels']
+  num_nodes = example['num_nodes']
+
+  senders = edge_index[:, 0]
+  receivers = edge_index[:, 1]
+
+  return jraph.GraphsTuple(
+      n_node=num_nodes,
+      n_edge=np.array([len(edge_index) * 2]),
+      nodes=node_feat,
+      edges=np.concatenate([edge_feat, edge_feat]),
+      # Make the edges bidirectional
+      senders=np.concatenate([senders, receivers]),
+      receivers=np.concatenate([receivers, senders]),
+      # Keep the labels with the graph for batching. They will be removed
+      # in the processed batch.
+      globals=np.expand_dims(labels, axis=0))
+
+
+def _get_weights_by_nan_and_padding(labels, padding_mask):
+  """Handles NaNs and padding in labels.
+
+  Sets all the weights from examples coming from padding to 0. Changes all NaNs
+  in labels to 0s and sets the corresponding per-label weight to 0.
+
+  Args:
+    labels: Labels including labels from padded examples
+    padding_mask: Binary array of which examples are padding
+  Returns:
+    tuple of (processed labels, corresponding weights)
+  """
+  nan_mask = np.isnan(labels)
+  replaced_labels = np.copy(labels)
+  np.place(replaced_labels, nan_mask, 0)
+
+  weights = 1.0 - nan_mask
+  # Weights for all labels of a padded element will be 0
+  weights = weights * padding_mask[:, None]
+  return replaced_labels, weights
+
+
+def _get_batch_iterator(dataset_iter, batch_size, num_shards=None):
+  """Turns a per-example iterator into a batched iterator.
+
+  Constructs the batch from num_shards smaller batches, so that we can easily
+  shard the batch to multiple devices during training. We use
+  dynamic batching, so we specify some max number of graphs/nodes/edges, add
+  as many graphs as we can, and then pad to the max values.
+
+  Args:
+    dataset_iter: The TFDS dataset iterator.
+    batch_size: How many average-sized graphs go into the batch.
+    num_shards: How many devices we should be able to shard the batch into.
+  Yields:
+    Batch in the init2winit format. Each field is a list of num_shards separate
+    smaller batches.
+  """
+  if not num_shards:
+    num_shards = jax.device_count()
+
+  # We will construct num_shards smaller batches and then put them together.
+  batch_size /= num_shards
```

Comment:
It looks like `batch_size` is here assumed to be the global batch size and not the per-device one?

Reply:
So given that we accumulate batches until we hit num_shards batches (`if count == num_shards:`), we should have been running with a per-device batch size of 256, for a total of 2048.

Reply:
Ah I see now, thanks!